General tablespace encryption

gl-sergei opened the pull request:
https://blueprints.launchpad.net/percona-server/+spec/encrypted-general-tablespaces

This patch does following:

1. Adds syntax `CREATE TABLESPACE ... ENCRYPTED='y/n'', this option then
   passed to InnoDB and added to tablespace flags. Since InnoDB
   encryption is really tablespace-based, not table-based, no further
   actions needed to encrypt the tablespace.
2. Validates encryption options for `CREATE TABLE' and `ALTER TABLE'
   statements

gl-sergei commented:
http://jenkins.percona.com/view/5.7/job/mysql-5.7-param/1027/

laurynas-biveinis commented on sql/handler.h:
```diff
@@ -623,6 +625,7 @@ class st_alter_tablespace : public Sql_alloc
     wait_until_completed= TRUE;
     ts_comment= NULL;
     ts_access_mode= TS_NOT_DEFINED;
+    encrypt= FALSE;
```
false

laurynas-biveinis commented on sql/sql_yacc.yy:
```diff
@@ -5044,6 +5045,14 @@ opt_ts_file_block_size:
+opt_ts_encryption:
+          ENCRYPTION_SYM opt_equal TEXT_STRING_sys
+	  {
+            Lex->alter_tablespace_info->encrypt= TRUE;
```
true

laurynas-biveinis commented on storage/innobase/fsp/fsp0fsp.cc:
```diff
@@ -278,9 +278,9 @@ fsp_flags_is_valid(
-	/* Only single-table and not temp tablespaces use the encryption
-	clause. */
-	if (is_encryption && (is_shared || is_temp)) {
+	/* Only single-table, general and not temp tablespaces use the
```
"Temporary tablespaces do not use the encryption clause"

laurynas-biveinis commented on storage/innobase/handler/ha_innodb.h:
```diff
@@ -756,6 +756,9 @@ class create_table_info_t
 	bool create_option_compression_is_valid();
 
+	/** Validate ENCRYPTION option. */
+	bool create_option_encryption_is_valid();
```
can it be made const?

gl-sergei commented:
All fixed.
http://jenkins.percona.com/view/5.7/job/mysql-5.7-param/1041/

percona-ysorokin commented on sql/handler.h:
```diff
@@ -623,6 +625,7 @@ class st_alter_tablespace : public Sql_alloc
     wait_until_completed= TRUE;
     ts_comment= NULL;
     ts_access_mode= TS_NOT_DEFINED;
+    encrypt= false;
```
`encrypt_type` must also be initialized here

gl-sergei replied:
fixed

percona-ysorokin commented on sql/sql_yacc.yy:
```diff
@@ -5044,6 +5045,14 @@ opt_ts_file_block_size:
         ;
 
+opt_ts_encryption:
+          ENCRYPTION_SYM opt_equal TEXT_STRING_sys
```
indentation

gl-sergei replied:
fixed

percona-ysorokin commented on storage/innobase/handler/ha_innodb.cc:
```diff
@@ -11860,18 +11919,6 @@ create_table_info_t::create_options_are_invalid()
 	}
 
-	/* Check the encryption option. */
-	if (ret == NULL && m_create_info->encrypt_type.length > 0) {
```
Any reason why `if (!create_option_encryption_is_valid())` is checked not here but much earlier?

gl-sergei replied:
Yes. I actually added comment for it. "Validate encryption
parameter even if `strict_mode` is OFF." Invalid compression option
with `strict_mode=OFF` does not make an error, while invalid
encryption option does. It was original documented MySQL behaviour
and I kept it. Fun part is that invalid encryption in vanilla server
fails in other place, not here :)

gl-sergei replied:
And there are also some lines in the test case checking this behaviour.

percona-ysorokin replied:
OK

percona-ysorokin commented on mysql-test/suite/innodb/t/ts_encrypt.test:
```diff
@@ -0,0 +1,174 @@
+--error ER_ILLEGAL_HA_CREATE_OPTION
+ALTER TABLE t6 TABLESPACE ts_encrypted;
+
+CREATE TABLESPACE ts_encrypted_new ADD DATAFILE 'ts_encrypted_new.ibd' ENCRYPTION="Y" ENGINE="InnoDB";
```
Please, also check adding second datafile to existing  encrypted / unencrypted tablespace specifying `ENCRYPTION="Y"/"N"`

gl-sergei replied:
this feature is not supported by InnoDB

percona-ysorokin commented on mysql-test/suite/innodb/t/ts_encrypt.test:
```diff
@@ -0,0 +1,174 @@
+call mtr.add_suppression("Partition.*does not exist in the InnoDB internal data dictionary though");
+
+--source include/have_innodb.inc
+--source include/not_embedded.inc
```
why `not_embedded`?

gl-sergei replied:
Does embedded support keyring plugins? I think I copied this one from table_encrypt_1 as it seemed logical to me

percona-ysorokin replied:
Oh, yes, you are right.

percona-ysorokin commented on mysql-test/suite/innodb/t/ts_encrypt.test:
```diff
@@ -0,0 +1,174 @@
+
+--source include/have_innodb.inc
+--source include/not_embedded.inc
+
```
please also add `have_partition.inc`

gl-sergei replied:
done

percona-ysorokin commented on mysql-test/suite/innodb/t/ts_encrypt.test:
```diff
@@ -0,0 +1,174 @@
+# SELECT FILE_ID, FILE_NAME FROM INFORMATION_SCHEMA.FILES
+# WHERE FILE_TYPE = 'TABLESPACE' ORDER BY FILE_ID;
+
+# SELECT NAME, FLAG, SPACE_TYPE FROM INFORMATION_SCHEMA.INNODB_SYS_TABLESPACES;
```
Add some `ORDER BY` clause.

gl-sergei replied:
this is just a commented-out example, but I added ORDER BY

percona-ysorokin commented on mysql-test/suite/innodb/t/ts_encrypt.test:
```diff
@@ -0,0 +1,174 @@
+DROP TABLE pt2;
+
+SELECT TABLE_NAME, PARTITION_NAME, SUBPARTITION_NAME, TABLESPACE_NAME
+FROM INFORMATION_SCHEMA.PARTITIONS WHERE TABLESPACE_NAME='ts_encrypted';
```
Although the result set should be empty, add `ORDER BY` here as well (just in case).

gl-sergei replied:
I just removed it because it was added to diagnose upstream bug 86809

percona-ysorokin commented on mysql-test/suite/innodb/t/ts_encrypt.test:
```diff
@@ -0,0 +1,174 @@
+# InnoDB general tablespace encryption
+
+call mtr.add_suppression("Partition.*does not exist in the InnoDB internal data dictionary though");
```
Please, add more comments to the test.
Just adding section names would make it much more readable. Say, "changing tablespace encryption options", "moving table to another tablespace", "partitions", etc.

percona-ysorokin commented on mysql-test/suite/innodb/t/ts_encrypt.test:
```diff
@@ -0,0 +1,174 @@
+CREATE TABLE t10 (a TEXT) ENGINE='InnoDB';
+CREATE TABLE t11 (a TEXT) ENCRYPTION='y' ENGINE='InnoDB';
+CREATE TABLE t12 (a TEXT) ENCRYPTION='n' ENGINE='InnoDB';
+
```
Please, also test "ALTER TABLE DISCARD/IMPORT TABLESPACE" for various encryption combinations.

gl-sergei replied:
InnoDB is able to export only file-per-table tablespaces

laurynas-biveinis commented:
@gl-sergei, is this currently ready for review?

gl-sergei commented:
Yes. It is.

laurynas-biveinis commented on mysql-test/suite/innodb/t/ts_encrypt.test:
```diff
@@ -0,0 +1,192 @@
+# InnoDB general tablespace encryption
```
Consider naming the testcase "general_ts_encrypt" or similar ("ts" here is not system, temporary, ...)

laurynas-biveinis commented on mysql-test/suite/innodb/t/ts_encrypt.test:
```diff
@@ -0,0 +1,192 @@
+# InnoDB general tablespace encryption
+
+call mtr.add_suppression("Partition.*does not exist in the InnoDB internal data dictionary though");
```
Suppression should be moved below the have_* not_* testcase skip checks

laurynas-biveinis commented on mysql-test/suite/innodb/t/ts_encrypt.test:
```diff
@@ -0,0 +1,192 @@
+ALTER TABLE t6 ENCRYPTION='Y';
+
+# moving tabbles from unencrypted tablespaces to encrypted ones and vice versa
+#  is not allowed
```
Extra space

laurynas-biveinis commented on mysql-test/suite/innodb/t/ts_encrypt.test:
```diff
@@ -0,0 +1,192 @@
+--error ER_ILLEGAL_HA_CREATE_OPTION
+ALTER TABLE t6 ENCRYPTION='Y';
+
+# moving tabbles from unencrypted tablespaces to encrypted ones and vice versa
```
s/tabbles/tables

laurynas-biveinis commented on mysql-test/suite/innodb/t/ts_encrypt.test:
```diff
@@ -0,0 +1,192 @@
+--error ER_ILLEGAL_HA_CREATE_OPTION
+ALTER TABLE t6 TABLESPACE ts_encrypted;
+
+# moving tabbles between tablespaces with the same encryption option is fine
```
s/tabbles/tables

laurynas-biveinis commented on mysql-test/suite/innodb/t/ts_encrypt.test:
```diff
@@ -0,0 +1,192 @@
+
+call mtr.add_suppression("Partition.*does not exist in the InnoDB internal data dictionary though");
+
+--source include/have_innodb.inc
```
--source include/have_partition.inc

gl-sergei replied:
its in line 7 :)

gl-sergei replied:
its in line 7 :)

laurynas-biveinis replied:
oh.

laurynas-biveinis commented on mysql-test/suite/innodb/t/ts_encrypt.test:
```diff
@@ -0,0 +1,192 @@
+CREATE TABLE t12 (a TEXT) ENCRYPTION='n' ENGINE='InnoDB';
+
+# test various combinations of partitioned tables with partitions created in
+# different tablespaces
```
Please expand partitioning tests to have subpartitions too

laurynas-biveinis commented on mysql-test/suite/innodb/t/ts_encrypt.test:
```diff
@@ -0,0 +1,192 @@
+SET GLOBAL innodb_file_per_table=OFF;
+SET GLOBAL innodb_strict_mode=OFF;
+
+# innodb_strict_mode change to take effect
```
Why not just set innodb_strict_mode session variable?

laurynas-biveinis commented on sql/sql_yacc.yy:
```diff
@@ -4816,6 +4816,7 @@ tablespace_option:
         | ts_wait
         | opt_ts_comment
         | opt_ts_file_block_size
+        | opt_ts_encryption
```
Create a new testcase in main suite for parser, this opt_ts_encryption clause. It should not require keyring options

laurynas-biveinis commented on mysql-test/suite/innodb/t/ts_encrypt.test:
```diff
@@ -0,0 +1,192 @@
+
+# check compatibility of different combinations of TABLESPACE and TABLE
+# encryption options
+
```
Add tests for invalid ENCRYPTION="blahblahblah", ENCRYPTION="" option values

gl-sergei commented:
PR updated with following changes:
- rename ts_encrypt -> general_ts_encrypt
- move suppression below have_*.inc
- add tests for invalid ENCRYPTION option values
- fix typo in "tabbles"
- removed extra space in comment
- expand partitioning tests to have subpartitions too
- replaced set global for `innodb_strict_mode` with set session
- create a new testcase in main suite for parser

http://jenkins.percona.com/view/5.7/job/mysql-5.7-param/1176/

gl-sergei commented:
Rebased on 5.7 trunk. Fixed one test failure after update. Updated param build: https://github.com/percona/percona-server/pull/1808

gl-sergei commented:
http://jenkins.percona.com/view/5.7/job/mysql-5.7-param/1180/

laurynas-biveinis commented on mysql-test/suite/innodb/t/general_ts_encrypt.test:
```diff
@@ -0,0 +1,266 @@
+ALTER TABLE q1 ENCRYPTION='Y',ALGORITHM=COPY;
+
+eval SET GLOBAL innodb_file_per_table=$innodb_file_per_table;
+eval SET SESSION innodb_strict_mode=$innodb_strict_mode;
```
Aren't you setting a session variable to saved global variable here? In any case there is no need to cleanup session variables

laurynas-biveinis commented on mysql-test/suite/innodb/t/general_ts_encrypt.test:
```diff
@@ -0,0 +1,266 @@
+call mtr.add_suppression("Partition.*does not exist in the InnoDB internal data dictionary though");
+
+let $innodb_file_per_table = `SELECT @@innodb_file_per_table`;
+let $innodb_strict_mode = `SELECT @@innodb_strict_mode`;
```
Should be redundant, see the comment at the end of testcase

gl-sergei replied:
Yes. Indeed. Removed this line and line 243 and re-recorded test result.
